File: python_leetcode/TwoSum.py
from typing import List


class Solution(object):
    def twoSum(self, nums: List, target: int) -> List:
        """Algorithm to find indexes of two numbers whose sum equals to the target number.
        Youtube explanation : https://www.youtube.com/watch?v=KLlXCFG5TnA

        Args:
            nums (List[int]): array containing numbers
            target (int): target number

        Returns:
            List: containing the indexes
        """

        prevMap = {}  # val -> index

        for i, n in enumerate(nums):
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i

File: java_leetcode/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;

public class TwoSum {

    public static void main(String[] args) {

        int[] testArray = {2, 7, 11, 15};
        int testTarget = 9;
        System.out.println(Arrays.toString(twoSum(testArray, testTarget)));

    }

    /**
     * Function to find indexes of two numbers in an array which add up to the given target number
     *
     * @param nums   int[]
     * @param target int
     * @return int[] , null if sum not found
     */
    public static int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> hashmap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {

            int diff = target - nums[i];

            if (hashmap.containsKey(diff)) {
                return new int[]{hashmap.get(diff), i};
            }
            hashmap.put(nums[i], i);
        }
        return null;
    }
}
